src/shared_utilities/implicit_timestepping.jl

```diff
@@ -107,53 +108,63 @@ to the `explicit_names` tuple.
-    fill!(parent(tridiag_field), NaN)
-
     # Only add jacobian blocks for fields that are in Y for this model
-    is_in_Y(name) = MatrixFields.has_field(Y, name)
+    is_in_Y(var) = MatrixFields.has_field(Y, var)
 
     # Define the implicit and explicit variables of any model we use
-    implicit_names = (@name(soil.ϑ_l), @name(soil.ρe_int))
-    explicit_names = (
+    implicit_vars =
+        (@name(soil.ϑ_l), @name(soil.ρe_int), @name(canopy.energy.T))
+    explicit_vars = (
         @name(soilco2.C),
         @name(soil.θ_i),
         @name(canopy.hydraulics.ϑ_l),
-        @name(canopy.energy.T),
         @name(snow.S),
         @name(snow.U)
     )
 
     # Filter out the variables that are not in this model's state, `Y`
-    available_implicit_names =
-        MatrixFields.unrolled_filter(is_in_Y, implicit_names)
-    available_explicit_names =
-        MatrixFields.unrolled_filter(is_in_Y, explicit_names)
+    available_implicit_vars =
+        MatrixFields.unrolled_filter(is_in_Y, implicit_vars)
+    available_explicit_vars =
+        MatrixFields.unrolled_filter(is_in_Y, explicit_vars)
+
+    get_jac_type(
+        space::Union{
+            Spaces.FiniteDifferenceSpace,
+            Spaces.ExtrudedFiniteDifferenceSpace,
+        },
+        FT,
+    ) = MatrixFields.TridiagonalMatrixRow{FT}
+    get_jac_type(
+        space::Union{Spaces.PointSpace, Spaces.SpectralElementSpace2D},
+        FT,
+    ) = MatrixFields.DiagonalMatrixRow{FT}
+
+    function get_j_field(space, FT)
+        jac_type = get_jac_type(space, FT)
+        # Create a field containing a `TridiagonalMatrixRow` at each point
+        field = ClimaCore.Fields.Field(jac_type, space)
+        fill!(parent(field), 0)
+        return field
+    end
 
+    implicit_blocks = MatrixFields.unrolled_map(
+        var ->
+            (var, var) =>
+                get_j_field(axes(MatrixFields.get_field(Y, var)), FT),
+        available_implicit_vars,
+    )
     # For explicitly-stepped variables, use the negative identity matrix
     # Note: We have to use FT(-1) * I instead of -I because inv(-1) == -1.0,
     # which means that multiplying inv(-1) by a Float32 will yield a Float64.
-    identity_blocks = MatrixFields.unrolled_map(
-        name -> (name, name) => FT(-1) * I,
-        available_explicit_names,
+    explicit_blocks = MatrixFields.unrolled_map(
+        var -> (var, var) => FT(-1) * I,
+        available_explicit_vars,
     )
 
-    # For implicitly-stepped variables, use a tridiagonal matrix
-    tridiagonal_blocks = MatrixFields.unrolled_map(
-        name ->
-            (name, name) => Fields.Field(
-                tridiag_type,
-                axes(MatrixFields.get_field(Y, name)),
-            ),
-        available_implicit_names,
-    )
 
-    matrix = MatrixFields.FieldMatrix(identity_blocks..., tridiagonal_blocks...)
+
+    matrix = MatrixFields.FieldMatrix(implicit_blocks..., explicit_blocks...)
```

Comment:
Before, we were setting up the FieldMatrix with the explicit (identity) blocks first, and then the implicit (tridiagonal). Do we need to switch the argument order here to `(explicit_blocks..., implicit_blocks...)`?

Reply:
RichardsModel and EnergyHydrology results look the same between this branch and main, so the order here probably doesn't matter (maybe it would if we had off-diagonal interaction terms)

Reply:
The FieldMatrix is equivalent to a Dict, so order doesn't matter here.